File: EHaskins.Frc.Dashboard/VirtualRobot/VirtualRobot.vb
Imports System.Net.Sockets
Imports System.Net

Public Class VirtualRobot
    Dim _isOpen As Boolean

    Dim _controlData As ControlData
    Dim _status As RobotStatus

    Dim _transmitClient As UdpClient
    Dim _receviceClient As UdpClient
    Dim _isConnected As Boolean = False

    Private _invalidPacketCount As Integer = 0
    Private _missedPacketCount As Integer = 0
    Private _packetCount As Integer = 0
    Private _teamNumber As Integer = 1103

    Public Sub New(ByVal teamNumber As Integer)
        Start(teamNumber)
    End Sub

    Public Event NewDataReceived As EventHandler

    Private Sub Start(ByVal teamNumber As Integer)
        _status = New RobotStatus()

        Me.TeamNumber = teamNumber

        _status.TeamNumber = Me.TeamNumber

        _transmitClient = New UdpClient()

        _receviceClient = New UdpClient(Configuration.DsToRobotLocalPortNumber)
        _isOpen = True
        _receviceClient.BeginReceive(AddressOf Me.ReceiveData, Nothing)
    End Sub

    Private Sub SendReply(ByVal packet As ControlData, ByVal endpoint As EndPoint)
        If _status.ReplyId > packet.PacketId And Not packet.ControlData.Resync Then
            Return
        End If
        _status.ReplyId = packet.PacketId
        _status.ControlData = packet.ControlData.Clone()

        Dim sendData = _status.CreateStatusPacket()
        Dim ipep As IPEndPoint = DirectCast(endpoint, IPEndPoint)

        ipep.Port = 1150

        _transmitClient.Send(sendData, sendData.Length, ipep)
    End Sub

    Public Sub ReceiveData(ByVal ar As IAsyncResult)
        Try
            Dim endpoint As IPEndPoint = Nothing
            Dim bytes = _receviceClient.EndReceive(ar, endpoint)
            Dim packet = ParseBytes(bytes)
            If packet IsNot Nothing AndAlso packet.IsValid AndAlso packet.TeamNumber = Me.TeamNumber Then
                SendReply(packet, endpoint)
                RaiseEvent NewDataReceived(Me, Nothing)
            Else
                InvalidPacketCount += 1
            End If
        Catch ex As Exception
            Throw
        Finally
            If _isOpen Then
                _receviceClient.BeginReceive(AddressOf Me.ReceiveData, Nothing)
            End If
        End Try
    End Sub


    Private Function ParseBytes(ByVal data As Byte()) As ControlData
        Try
            _packetCount += 1
            Dim packet = New ControlData(data)
            ControlData = packet

            Debug.WriteLine(ControlData.PacketId)

            If Not ControlData.IsValid Then
                InvalidPacketCount += 1
            End If

            Return packet
        Catch ex As Exception
            Debug.WriteLine(ex.Message)
            Throw
        End Try
    End Function

    Public Property IsConnected() As Boolean
        Get
            Return _isConnected
        End Get
        Set(ByVal value As Boolean)
            _isConnected = value
        End Set
    End Property
    Public Property ControlData() As ControlData
        Get
            Return _controlData
        End Get
        Private Set(ByVal value As ControlData)
            _controlData = value
        End Set
    End Property
    Public ReadOnly Property Status() As RobotStatus
        Get
            Return _status
        End Get
    End Property

    Public Property InvalidPacketCount() As Integer
        Get
            Return _invalidPacketCount
        End Get
        Set(ByVal value As Integer)
            _invalidPacketCount = value
        End Set
    End Property

    Public Property TeamNumber() As Integer
        Get
            Return _teamNumber
        End Get
        Set(ByVal value As Integer)
            _teamNumber = value
        End Set
    End Property

End Class

File: EHaskins.Frc.Communication/VirtualRobot.cs
using System;
using System.Net.Sockets;
using System.Net;
using MicroLibrary;
using System.Diagnostics;

namespace EHaskins.Frc.Communication
{
    public class VirtualRobot
    {
        bool _isOpen;

        StatusData _status;
        UdpClient _transmitClient;
        UdpClient _receviceClient;

        bool _isConnected = false;
        private int _invalidPacketCount = 0;
        private int _packetCount = 0;

        private ushort _teamNumber = 1103;
        public VirtualRobot(ushort teamNumber)
        {
            this.TeamNumber = teamNumber;

            TransmitPort = Configuration.RobotToDsDestinationPortNumber;
            ReceivePort = Configuration.DsToRobotDestinationPortNumber;
            UserControlDataLength = Configuration.UserControlDataSize;
            UserStatusDataLength = Configuration.UserStatusDataSize;
        }

        public event EventHandler NewDataReceived;
        public int TransmitPort { get; set; }
        public int ReceivePort { get; set; }
        public int UserStatusDataLength { get; set; }
        public int UserControlDataLength { get; set; }
        public bool IsConnected
        {
            get { return _isConnected; }
            set { _isConnected = value; }
        }
        public ControlData ControlData { get; set; }
        public StatusData StatusData
        {
            get { return _status; }
        }
        public int InvalidPacketCount
        {
            get { return _invalidPacketCount; }
            set { _invalidPacketCount = value; }
        }
        public ushort TeamNumber
        {
            get { return _teamNumber; }
            set { _teamNumber = value; }
        }


        public void Start()
        {
            Console.WriteLine("Starting team " + TeamNumber + "on receive:" + ReceivePort + ", transmit:" + TransmitPort);
            _status = new StatusData();
            _status.UserStatusDataLength = UserControlDataLength;

            _status.TeamNumber = this.TeamNumber;

            ControlData = new ControlData();

            _transmitClient = new UdpClient();

            _receviceClient = new UdpClient(ReceivePort);
            _isOpen = true;
            _receviceClient.BeginReceive(this.ReceiveData, null);
        }

        private void SendReply(ControlData packet, EndPoint endpoint)
        {
            if (_status.ReplyId > packet.PacketId & !packet.Mode.IsResync)
            {
                IsConnected = false;
                return;
            }
            _status.ReplyId = packet.PacketId;
            //_status.ControlData = new ControlData(); // packet.Mode.Clone();//TODO:FIX
            //TODO:FIX
            var sendData = _status.GetBytes() 
            IPEndPoint ipep = (IPEndPoint)endpoint;

            ipep.Port = TransmitPort;

            _transmitClient.Send(sendData, sendData.Length, ipep);
        }

        public void ReceiveData(IAsyncResult ar)
        {
            try
            {
                IPEndPoint endpoint = null;
                var bytes = _receviceClient.EndReceive(ar, ref endpoint);
                bool lastEStop = ControlData.Mode.IsEStop;
                if (bytes.IsValidFrcPacket())
                    ParseBytes(bytes);

                if (ControlData != null && ControlData.TeamNumber == this.TeamNumber)
                {
                    SendReply(ControlData, endpoint);
                    if (ControlData != null && lastEStop)
                    {
                        ControlData.Mode.IsEStop = true;
                    }
                    if (NewDataReceived != null)
                    {
                        NewDataReceived(this, null);
                    }
                }
                else
                {
                    InvalidPacketCount += 1;
                }
            }
            catch (Exception ex)
            {
                throw;
            }
            finally
            {
                if (_isOpen)
                {
                    _receviceClient.BeginReceive(this.ReceiveData, null);
                }
            }
        }

        private void ParseBytes(byte[] data)
        {
            try
            {
                _packetCount += 1;
                    ControlData.Update(data);
            }
            catch (Exception ex)
            {
                Debug.WriteLine(ex.Message);
                throw;
            }
        }
    }
}
